client.go

```diff
@@ -0,0 +1,436 @@
+package client
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+)
+
+var emailValid = regexp.MustCompile(`^[a-z0-9._%+\-]+@[a-z0-9.\-]+\.[a-z]{2,4}$`)
+
+// InvalidEmail Error
+type InvalidEmail struct{}
+
+func (e InvalidEmail) Error() string {
+	return "Invalid Email format"
+}
+
+// RoleMissing Error
+type RoleMissing struct{}
+
+func (e RoleMissing) Error() string {
+	return "Missing Role Parameter"
+}
+
+// NameMissing Error
+type NameMissing struct{}
+
+func (e NameMissing) Error() string {
+	return "Missing Name Parameter"
+}
+
+// EmailMissing Error
+type EmailMissing struct{}
+
+func (e EmailMissing) Error() string {
+	return "Missing Email Parameter"
+}
+
+// LoginMissing Error
+type LoginMissing struct{}
+
+func (e LoginMissing) Error() string {
+	return "Missing Login Parameter"
+}
+
+// PasswordMissing Error
+type PasswordMissing struct{}
+
+func (e PasswordMissing) Error() string {
+	return "Missing Password Parameter"
+}
+
+// NotInteger Error
+type NotInteger struct{}
+
+func (e NotInteger) Error() string {
+	return "ID provided is not Integer"
+}
+
+// TokenExpired Error
+type TokenExpired struct {
+}
+
+func (e TokenExpired) Error() string {
+	return "Expired Token"
+}
+
+// NotFound Error
+type NotFound struct {
+}
+
+func (e NotFound) Error() string {
+	return "Token Not Found"
+}
+
+// VisualizationError errors for Visualization client
+type VisualizationError struct {
+	code        string
+	message     string
+	description string
+}
+
+// Error generate a error message.
+// If Code is zero, we know it's not a http error.
+func (e VisualizationError) Error() string {
+	return fmt.Sprintf("ERROR: %s", e.description)
+}
+
+// VisualizationClient client for Visualization
+type VisualizationClient struct {
+	url            string
+	client         *http.Client
+	token          AuthToken
+	openstackToken string
+}
+
+// NewVisualizationClient returns client with token
+func NewVisualizationClient(url string, token AuthToken) (*VisualizationClient, error) {
+	client := http.Client{}
+	return &VisualizationClient{client: &client, url: url, token: token}, nil
+
+}
+
+// A JSONMessage contains the json error message received when http request failed
+type JSONMessage struct {
+	Message string `json:"message"`
+}
+
+// httpRequest handle the request to Grafana server.
+//It returns the response body and a error if something went wrong
+func (v *VisualizationClient) httpRequest(method string, url string, body io.Reader) (result io.Reader, err error) {
+
+	// validate token
+	if v.token == (AuthToken{}) {
```

Comment:
Move token validation logic one level up.
f.e.
- GetUsers // json models here
   - authorizedHttpRequest // token checking and reissuing new one here
      - httpRequest // just http call here